Add Arduino Portenta H7 as target

facchinm opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

This PR series introduces the support for [Arduino Portenta H7](https://store.arduino.cc/portenta-h7) .
The patchset has been split into independent PRs for an easier review.
#13816 
#13817 
~#13863~

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

/cc @pennam @pnndra

#### Impact of changes <!-- Optional -->

Except the standalone `TARGET_PORTENTA*` content, some other files have been touched (all of them in separate commits).

In particular, `COMPONENT_WHD` has been made generic (any mbed target implementing `SERIAL_FC` can now compile it) but this could impact PSOC6 targets which used to configure it via Cypress HAL.

<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR

Test results attached (this branch build over  #13863 )

[report-13-11.zip](https://github.com/ARMmbed/mbed-os/files/5535651/report-13-11.zip)

    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

@jeromecoutant 
@romanjoe (for the WHD patches)

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

facchinm committed:
STM32H7: dual core: handle interrupts on CM4

facchinm committed:
STM32: USBPhy: enable HS clock sleep

facchinm committed:
STM32H747xI: allow overriding SetSysClock

facchinm committed:
Add Arduino Portenta H7 target

facchinm committed:
lwip: use adeguate mem-size for Portenta H7

facchinm committed:
STM32: EMAC: allow compiling on MCUs without DCACHE

facchinm committed:
PortentaH7: add Ethernet support

facchinm committed:
STM32H7: DualCore: allow no forced CM4 boot

facchinm committed:
STM32: QSPI: allow setting SAMPLE_SHIFTING mode

facchinm committed:
Fix missing licenses for Arduino original code

jeromecoutant commented on targets/targets.json:
```diff
@@ -2836,6 +2836,111 @@
+    "PORTENTA_H7": {
+        "core": "",
+        "public": false,
+        "inherits": ["MCU_STM32"],
```
Did you check MCU_STM32H7, and MCU_STM32H747xI_CM7/CM4 ?

facchinm replied:
The commit was done before the unification of `MCU_STM32*` targets, so it can probably be rewritten in a "modern" way before merging. Thanks for pointing it out

facchinm replied:
Done!

jeromecoutant commented on targets/TARGET_STM/rtc_api.c:
```diff
@@ -65,7 +65,11 @@ void rtc_init(void)
     RCC_OscInitStruct.PLL.PLLState   = RCC_PLL_NONE;
+#if MBED_CONF_TARGET_LSE_BYPASS
+    RCC_OscInitStruct.LSEState       = RCC_LSE_BYPASS;
+#else
```
Should be removed as part of #13817 ?

facchinm replied:
Yup, I'll rebase today since the other patch was merged

jeromecoutant commented on targets/TARGET_STM/gpio_irq_api.c:
```diff
@@ -149,9 +149,13 @@ static void handle_interrupt_in(uint32_t irq_index, uint32_t max_num_pin_line)
 #else /* TARGET_STM32L5 */
 
             // Clear interrupt flag
+#if defined(DUAL_CORE) && defined(CORE_CM4)
+            if (__HAL_GPIO_EXTID2_GET_FLAG(pin) != RESET) {
+                __HAL_GPIO_EXTID2_CLEAR_FLAG(pin);
+#else
```
@ABOSTM Could you have a look on this ?

facchinm replied:
A bit more information and code to reproduce the problem can be found here https://github.com/arduino/ArduinoCore-mbed/pull/68#issue-499769014 .

ABOSTM replied:
Yes it looks good to distinguish between interrupts for Core1 and for Core2.
But I wonder if it a good chose to use CORE_CM4, because it is right that on H7  CM4 = Core2, but not sure it will be valid for further targets

jeromecoutant replied:
Let's keep CORE_CM4

jeromecoutant commented on targets/TARGET_STM/USBPhy_STM32.cpp:
```diff
@@ -29,14 +29,29 @@
-#define NUM_ENDPOINTS           4
+
+#if (MBED_CONF_TARGET_USB_SPEED == USE_USB_OTG_HS)
+
```
part of #13816

facchinm replied:
Same here, will rebase after #13816 gets merged

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_PORTENTA_H7/PeripheralNames.h:
```diff
@@ -0,0 +1,106 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *******************************************************************************
+ */
+#ifndef MBED_PERIPHERALNAMES_H
```
You should remove this file, as it is duplicated:
https://github.com/ARMmbed/mbed-os/blob/master/targets/TARGET_STM/TARGET_STM32H7/PeripheralNames.h

facchinm replied:
Indeed, thanks, I'll remove it

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_PORTENTA_H7/COMPONENT_WHD/mbed_lib.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "name": "cy_psoc6_whd",
```
Is it the right place to add this COMPONENT ?
Maybe could be better in connectivity directory ?

facchinm replied:
Making the whole WHD driver completely HW independent is not a super easy task (and needs coordination with Cypress folks). In the meantime I preferred to keep it internal so it's clear that it needs further rework to be used with other boards)

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_PORTENTA_H7/system_clock_override.c:
```diff
@@ -0,0 +1,304 @@
+    HAL_GPIO_WritePin(GPIOH, GPIO_PIN_1, 1);
+
+    /* Supply configuration update enable */
+#if HSE_VALUE == 27000000
```
Why this value ?
I can see value is set to 25000000 in targets.json

facchinm replied:
The oscillator on beta version of the boards was 27Mhz; we used to recognize it at runtime using a flash-based information "table" but the combination of no OTP area, big erase pages on H7 and the specific use case made me think it was better to keep as a non-default configuration value. 
It can be safely removed for the sake of clarity :slightly_smiling_face:

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/system_clock.c:
```diff
@@ -160,7 +160,7 @@ MBED_WEAK uint8_t SetSysClock_PLL_HSE(uint8_t bypass)
 /*            PLL (clocked by HSI) used as System clock source                */
 /******************************************************************************/
-uint8_t SetSysClock_PLL_HSI(void)
+MBED_WEAK uint8_t SetSysClock_PLL_HSI(void)
```
You should not have to update this function.
I can see the difference is HAL_PWREx_ConfigSupply call, I will think about this...

facchinm replied:
Yup, we need a way to override the power supply scheme and it was the easiest one

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_PORTENTA_H7/system_clock_override.c:
```diff
@@ -0,0 +1,304 @@
+/******************************************************************************/
+/*            PLL (clocked by HSE) used as System clock source                */
+/******************************************************************************/
+uint8_t SetSysClock_PLL_HSE(uint8_t bypass, bool lowspeed)
```
Could you add some comment about this 2 speed mode ?
Thx

facchinm replied:
Sure, will do

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_PORTENTA_H7/system_clock_override.c:
```diff
@@ -0,0 +1,304 @@
+}
+#endif /* ((CLOCK_SOURCE) & USE_PLL_HSI) */
+
+#if defined (CORE_CM4)
+void HSEM2_IRQHandler(void)
+{
+  HAL_HSEM_IRQHandler();
+}
+#endif
+
+#if defined (CORE_CM7)
+void HSEM1_IRQHandler(void)
+{
+  HAL_HSEM_IRQHandler();
+}
+#endif
```
@ABOSTM Should we add this in default H7 dual core configuration ?

ABOSTM replied:
No. From my point of view, this code is useless even for Portenta targets, because we don't use HSEM interrupts, except for dual core boot up synchronisation, and in this case we just use the wakeup mechanism, not the nvic callback. 
So this code should be removed.

facchinm replied:
We use that handlers for [RPC](https://github.com/arduino/ArduinoCore-mbed/tree/master/libraries/RPC) library. I agree that these lines can be moved there or somewhere else for clarity.

ABOSTM replied:
OK then it is needed only for  Portenta targets

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_PORTENTA_H7/PeripheralPins.c:
```diff
@@ -0,0 +1,584 @@
+
+MBED_WEAK const PinMap PinMap_ADC[] = {
+    {PA_0,       ADC_1, STM_PIN_DATA_EXT(STM_MODE_ANALOG, GPIO_NOPULL, 0, 16, 0)}, // ADC1_INP16
+    {PA_0C,      ADC_1, STM_PIN_DATA_EXT(STM_MODE_ANALOG, GPIO_NOPULL, 0, 0, 0)}, // ADC1_INP0 // Connected to ARD_A2
```
I think all comments are not really correct ?

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_PORTENTA_H7/PeripheralPins.c:
```diff
@@ -0,0 +1,584 @@
+MBED_WEAK const PinMap PinMap_ADC_Internal[] = {
+    {ADC_TEMP,   ADC_1,    STM_PIN_DATA_EXT(STM_MODE_ANALOG, GPIO_NOPULL, 0, 16, 0)},
+    {ADC_VREF,   ADC_1,    STM_PIN_DATA_EXT(STM_MODE_ANALOG, GPIO_NOPULL, 0, 17, 0)},
+    {ADC_VBAT,   ADC_1,    STM_PIN_DATA_EXT(STM_MODE_ANALOG, GPIO_NOPULL, 0, 18, 0)},
```
You could copy from DISCO_H747I file the correct configuration

ABOSTM commented on targets/TARGET_STM/gpio_irq_api.c:
```diff
@@ -501,6 +505,10 @@ void gpio_irq_enable(gpio_irq_t *obj)
+#if defined(DUAL_CORE) && defined(CORE_CM4)
+    LL_C2_EXTI_EnableIT_0_31(1 << pin_index);
+#endif
+
```
In case of dualcore CM4, I think we should enabe only CM4 interrupt
```suggestion
#if defined(DUAL_CORE) && defined(CORE_CM4)
    LL_C2_EXTI_EnableIT_0_31(1 << pin_index);
#else
    LL_EXTI_EnableIT_0_31(1 << pin_index);
#endif

```

ABOSTM commented on targets/TARGET_STM/gpio_irq_api.c:
```diff
@@ -525,6 +533,10 @@ void gpio_irq_disable(gpio_irq_t *obj)
 
+#if defined(DUAL_CORE) && defined(CORE_CM4)
+    LL_C2_EXTI_DisableIT_0_31(1 << pin_index);
+#endif
```
Similarly disable only CM4 interrupt in case of dualcore CM4

jeromecoutant commented on targets/TARGET_STM/mbed_overrides.c:
```diff
@@ -129,11 +129,14 @@ void mbed_sdk_init()
         LL_HSEM_ReleaseLock(HSEM, CFG_HW_STOP_MODE_SEMID, 0);
-    } else {
+    }
+#ifndef NO_FORCED_CM4_BOOT
```
Do you have a specific Option Byte configuration ?

facchinm replied:
Yes, our configuration is SYSCFG_UR1_BCM4 = 0 but we don't want to force the CM4 boot (because it could not contain any valid binary)

ABOSTM replied:
If CM4 doesn't contains valid binary then switch `DUAL_CORE `should be disabled, and thus no need for `NO_FORCED_CM4_BOOT`

jeromecoutant replied:
Not so easy as DUAL_CORE is define at CMSIS level in stm32h747xx.h

ABOSTM replied:
@jeromecoutant , is there no mbed switch to define dual core usage ?

ABOSTM replied:
@facchinm, after discussion with @jeromecoutant:
As there is currently no real dual core management in mbed, we agree to add this new compialtion switch NO_FORCED_CM4_BOOT
Nevertheless, in order to have the same behavior (like your example where CM4 runs only when RPC lib requires it ), whatever OptionByte BOOT_CM4, we propose to move `#ifndef NO_FORCED_CM4_BOOT` few lines above, just before the`if` so that both `if `and `else `cases are covered

```
#ifndef NO_FORCED_CM4_BOOT
   if (READ_BIT(SYSCFG->UR1, SYSCFG_UR1_BCM4)) {
        LL_HSEM_1StepLock(HSEM, CFG_HW_STOP_MODE_SEMID);
        /*Release HSEM in order to notify the CPU2(CM4)*/
        LL_HSEM_ReleaseLock(HSEM, CFG_HW_STOP_MODE_SEMID, 0);
    }
```

jeromecoutant replied:
Macro name could be, then, something like `CM4_BOOT_BY_APPLICATION` ?

facchinm replied:
@jeromecoutant @ABOSTM I agree on changing the define but I'm unsure about broadening the scope of the `ifndef`.
Let me expand:
if you leave BCM4 bit == 1 I presume you want to automatically boot the M4 too, so it makes sense to synchronize the processors (otherwise the M4 would boot with garbage in any case).
The other case is a bit more convoluted, so the user may or may not have code on the M4 and boot it via the application.

ABOSTM replied:
I understand your point but relying on Option bytes to adapt software behavior suppose Hardware settings, Option Bytes, are in line with Software application development, and are set on purpose.
That is probably the case when you develops your application, but there may be users that are not aware of their Option byte settings, or just keep default , or chnage then for test, debug accidentally, ...
Mbed is a generic framework, we don't know which application will be developed, we cannot guarrante OpionBytes settings. So I prefer to avoid relying on hypothetical alignment between hardware and software. Instead I prefer to have the same behavior whatever Option Bytes setting. This is simpler for beginner user, and may prevent misalignment problem. 
In other words, Software configuration (new switch CM4_BOOT_BY_APPLICATION) take the precedence on hardware settings (Option Bytes).

From my point of view, STM32H747 in Mbed today doesn't tackle the case of no binary in CM4, and even less the case where M4 binary is loaded after boot by CM7.  And I am not sure we want to go this way.

facchinm replied:
Agree! I reworked the commits to apply both the changes

ABOSTM commented on targets/TARGET_STM/mbed_overrides.c:
```diff
@@ -129,11 +129,14 @@ void mbed_sdk_init()
-    } else {
+    }
+#ifndef NO_FORCED_CM4_BOOT
+    else {
```
I feel like it is not needed:
Code is under DUAL_CORE switch. So CM4 need to boot somehow.
* Either it boots at reset in parallel of CM7(depending on Option bytes),
and in this case CM4 automatically go to sleep mode, waiting for CM7 to configure clock (because it has been decided in Mbed that only CM7 is able to set clock), Then CM7 need to wakeup CM4. This correspond to `if (READ_BIT(SYSCFG->UR1, SYSCFG_UR1_BCM4)`
* Or CM4 doesn't boot at reset (depending on Option bytes) and CM7 need to force CM4 boot. This is the `esle`

facchinm replied:
I don't understand why booting the CM4 should be compulsory :slightly_smiling_face: 
In fact, we are leaving the chance to boot it later when RPC lib requires it or to flash a CM7 only binary that acts as an H743 (more or less)

ABOSTM reviewed (approved):
LGTM for dual core part

jeromecoutant commented on connectivity/mbedtls/source/x509_crt.c:
```diff
@@ -80,7 +80,11 @@
+#include <platform/mbed_retarget.h>
+#else
 #include <dirent.h>
+#endif /* __MBED__ */
```
Should be in a specific PR ?

facchinm replied:
Probably, I inserted it here because 6a1c6b9058c3b51d6a3913f13dfb1c1133e26914 depends on it but we can spun it out

facchinm replied:
Standalone PR created https://github.com/ARMmbed/mbed-os/pull/13863

jeromecoutant commented on connectivity/netsocket/source/TLSSocketWrapper.cpp:
```diff
@@ -103,8 +103,17 @@ nsapi_error_t TLSSocketWrapper::set_root_ca_cert(const void *root_ca, size_t len
+        ret = mbedtls_x509_crt_parse(crt, static_cast<const unsigned char *>(root_ca), len);
+        break;
+    }
+    if (ret < 0) {
```
Should be in a specific PR ?

jeromecoutant reviewed (commented):
Will execute tests when I have HW!

pennam committed:
Portenta: enable ETH power supply

facchinm committed:
Add SPDX-License

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_PORTENTA_H7_M7/TOOLCHAIN_GCC_ARM/STM32H747xI.ld:
```diff
@@ -0,0 +1,227 @@
+        . += M_CRASH_DATA_RAM_SIZE;
+        . = ALIGN(8);
+        __CRASH_DATA_RAM_END__ = .; /* Define a global symbol at data end */
+    } > RAM
```
Note that you crash feature is not enabled for Portenta:
https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_lib.json#L129

facchinm replied:
You advise adding it in mbed_lib.json or at target.json level?

jeromecoutant replied:
or to remove the section.
I don't really know if crash feature is well supported and useful ?

facchinm replied:
Isn't it needed if the user overrides `crash-capture-enabled` in its app? I'd keep it, in the worst case it doesn't use any memory if the flag is off

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_PORTENTA_H7_M7/TOOLCHAIN_GCC_ARM/STM32H747xI.ld:
```diff
@@ -0,0 +1,227 @@
+
+#if !defined(RAM_SIZE)
+  #define RAM_SIZE 512K
+#endif
```
Note that this code is already available in ../cmsis_nvic.h

facchinm replied:
I'd love to switch and inherit from `TARGET_STM32H747xI_CM7`/`TARGET_STM32H747xI_CM4` but I doubt that the current infrastructure allows importing linker scripts over other linker scripts (we need a modification for openAMP to run and that's a mayor use case so I'd keep it at platform level)

jeromecoutant replied:
No problem,
just indicated that you could replace this part of code by #include "../cmsis_nvic.h"

facchinm replied:
Fixing right now, thanks for the hint

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_PORTENTA_H7/COMPONENT_WHD/generated_mac_address.txt:
```diff
@@ -0,0 +1,9 @@
+ * MAC address is printed during WHD power up
+ */
+
+#define NVRAM_GENERATED_MAC_ADDRESS        "macaddr=00:A0:50:6f:b2:ea"
```
Should this macro be define in some json file in order to be reconfigurable ?

facchinm replied:
That's just a placeholder (the real MAC address is written in the silicon) but that's the method Cypress uses so I kept is

jeromecoutant commented on targets/targets.json:
```diff
@@ -2885,6 +2885,86 @@
+            "CORDIO"
+        ],
+        "features": [
+            "LWIP",
```
I don't think LWIP features is needed ? (deprecated)

jeromecoutant commented on targets/targets.json:
```diff
@@ -2885,6 +2885,86 @@
+        ],
+        "device_has_add": [
+            "USBDEVICE",
+            "FLASH",
```
Already defined at MCU_STM32H7 level

jeromecoutant commented on targets/targets.json:
```diff
@@ -2885,6 +2885,86 @@
+        ],
+        "overrides": {
+            "clock_source": "USE_PLL_HSE_EXTC",
+            "lse_available": 0,
```
Sure ?

facchinm replied:
Yes, for compatibility with Beta boards (of course, we can maintain an out of tree patch for this too)

jeromecoutant commented on targets/targets.json:
```diff
@@ -2885,6 +2885,86 @@
+            "lpticker_delay_ticks": 0,
+            "network-default-interface-type": "ETHERNET"
+        },
+        "release_versions": ["5"],
```
I think this is deprecated info

jeromecoutant commented on targets/targets.json:
```diff
@@ -2885,6 +2885,86 @@
+        },
+        "release_versions": ["5"],
+        "device_name": "STM32H747XIHx",
+        "bootloader_supported": true
```
Already defined at MCU_STM32 level

jeromecoutant commented on targets/targets.json:
```diff
@@ -2885,6 +2885,86 @@
+        "mbed_ram_size" : "0x48000",
+        "macros_add": [
+            "CORE_CM4",
+            "MBED_MPU_CUSTOM"
```
Is it mandatory ?

facchinm replied:
It used to be in our mixed environment (micropython running on M7 and mbed/Arduino on the M4) ~not sure if it still applies~ 
I can confirm that it's still compulsory for our use case

jeromecoutant commented on targets/targets.json:
```diff
@@ -2885,6 +2885,86 @@
+            "MBED_MPU_CUSTOM"
+        ],
+        "device_has_remove": [
+            "MPU"
```
@ABOSTM

jeromecoutant reviewed (commented):
Only GCC is supported ?

Note I couldn't compile...
````
[Error] x509_crt.c@1603,57: dereferencing pointer to incomplete type 'struct dirent'
````

jeromecoutant commented on targets/targets.json:
```diff
@@ -2885,6 +2885,82 @@
+            "clock_source": "USE_PLL_HSE_EXTC",
+            "lse_available": 1,
+            "lpticker_delay_ticks": 0,
+            "network-default-interface-type": "ETHERNET"
```
I don't see any Ethernet on the board...
Should be WIFI ? or CELLULAR ?

facchinm replied:
The board has both an Ethernet phy mounted (exported like in the https://store.arduino.cc/portenta-vision-shield) and a WiFi module (but the fw lives in a user-selected filesystem so it's not useable "out of the box" in mbed).
I'd keep ETHERNET as default interface since it's zero config but no problem in changing it.

jeromecoutant replied:
Maybe several targets can be created depending on the several HW configuration that can exist ?

facchinm committed:
[Portenta] Add Cmake rules

Completes #13915

facchinm committed:
STM32H747_CM4: add RAM_D3 memory section

facchinm committed:
PORTENTA_H7: use default linker/startup scripts

hugueskamba reviewed (dismissed):
Thank you for your changes.
https://github.com/ARMmbed/mbed-os/blob/master/targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/CMakeLists.txt#L10 also needs to be modified to add the `TARGET_PORTENTA_H7` dir.
```cmake
...
elseif("PORTENTA_H7" IN_LIST MBED_TARGET_LABELS)
    add_subdirectory(TARGET_PORTENTA_H7)
elseif("STM32H747xI_CM4" IN_LIST MBED_TARGET_LABELS)
...
```

facchinm committed:
cmake: STM32H7: fix mbed-emac target

facchinm committed:
cmake: STM32H7: fix stm32h747 targets

This needs to be defined for "final" targets not contained in their processor subdirectory

facchinm committed:
cmake: STM32H7: enable PORTENTA* target

hugueskamba commented on targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_PORTENTA_H7/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) 2020 ARM Limited. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+target_sources(mbed-core
```
Can this just be moved to `targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/CMakeLists.txt` ?

facchinm replied:
Mmmh why should it be moved there? It's board specific so I'd leave it self contained

hugueskamba replied:
From the soon to be released Mbed [target porting guide](https://github.com/ARMmbed/mbed-os-5-docs/blob/development/docs/porting/target/cmake.md):
 >Add `CMakeLists.txt` files to the top level directory of a given vendor directory. List all files found in the directory in this CMake input source file, adding additional CMake input source file if it is MCU or Mbed target specific and has a great number of files which will make the top level `CMakeLists.txt` too complex. Think when you decide to create functions in a computer software code to remove complexity.
 

 
 However, as you said below it can be with an atomic patch to also fix the other Mbed targets.

hugueskamba commented on connectivity/drivers/emac/TARGET_STM/TARGET_STM32H7/TARGET_PORTENTA_H7/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+# Copyright (c) 2020 ARM Limited. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+target_sources(mbed-emac
```
As the content is too small, this can be added in the list file that add this one to the build.
This list file can then just be deleted.

facchinm replied:
The same applies to all files in that subdirectory, like `connectivity/drivers/emac/TARGET_STM/TARGET_STM32H7/TARGET_NUCLEO_H743ZI/CMakeLists.txt`. Can we do it with an atomic patch after the merge?
